sdk/go/common/resource/properties.go

```diff
@@ -646,3 +646,47 @@ const OutputValueSig = "d0e6a833031e9bbcd3f4e8bde6ca49a4"
 func IsInternalPropertyKey(key PropertyKey) bool {
 	return strings.HasPrefix(string(key), "__")
 }
+
+func (m *PropertyMap) Special() SpecialProperties {
+	return SpecialProperties{m}
+}
+
+type SpecialProperties struct {
+	m *PropertyMap
+}
+
+func (sp SpecialProperties) Version() SpecialProperty {
+	return SpecialProperty{
+		name: "version",
+		r:    sp.m,
+	}
+}
+
+func (sp SpecialProperties) ServerURL() SpecialProperty {
+	return SpecialProperty{
+		name: "pluginDownloadURL",
+		r:    sp.m,
+	}
+}
```

Comment:
I added these helpers because I wanted to distinguish between inputs that Pulumi (the engine) cares about and general inputs (that are user defined). It also provides a single source of truth for the correct key, which prevents mistypes. We can just look up the value if you feel that would be a better option.

Reply:
My preference is not expose the helpers here. `sdk/go/common/resource/properties` is used across the Pulumi ecosystem to represent properties for any resource, not just providers. Adding new public APIs here with special accessors for two provider inputs doesn't feel quite right to me.

I'd be fine with having non-public helpers that can be used within the implementation. Perhaps in pkg/internal? Or no helpers at all and just access the map directly (possibly using string constants if we're concerned about mistypes).

If you are going to keep some helpers like this, one nit (related to my other comment): The accessor looking up `"pluginDownloadURL"` should probably be named `PluginDownloadURL()`. I think it's better for the key and name of the accessor to match.

Reply:
That makes perfect sense. I'm going to move the helpers to `pkg/resource/deploy/providers`. There is already a helper  (`GetProviderVersion`), so the location should make sense. I'll make the name change here, and just make the general change from `ServerURL` to `PluginDownloadURL`.